File: backend/app/api/v1/member_query.py
"""Roster selection — the vocabulary and the headcount behind a member query.

These endpoints exist so a broker never has to run a preview (or type a staff id)
to find out who a rule matches:

- GET  /policy-years/{id}/member-facets        — filter vocabulary + headcounts
- POST /policy-years/{id}/member-query/count   — live "N members match"
- POST /policy-years/{id}/member-query/resolve — pasted list → members

They are deliberately independent of the bulk-update module: a selection is a
general capability, and Coverage & Members is the next surface that wants it.
"""
from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.deps import load_policy_year
from app.db.session import get_db
from app.models import PolicyYear
from app.schemas.member_query import (
    MemberFacetsOut,
    MemberListResolveIn,
    MemberListResolveOut,
    MemberQueryCountIn,
    MemberQueryCountOut,
)
from app.services import member_query as svc
from app.services.enrollment_products import resolve_product_by_code

router = APIRouter(tags=["member-query"])


@router.get(
    "/policy-years/{policy_year_id}/member-facets", response_model=MemberFacetsOut
)
def member_facets(
    policy_year_id: str,
    py: PolicyYear = Depends(load_policy_year),
    db: Session = Depends(get_db),
) -> MemberFacetsOut:
    return svc.build_facets(db, py)


@router.post(
    "/policy-years/{policy_year_id}/member-query/count",
    response_model=MemberQueryCountOut,
)
def member_query_count(
    policy_year_id: str,
    body: MemberQueryCountIn,
    py: PolicyYear = Depends(load_policy_year),
    db: Session = Depends(get_db),
) -> MemberQueryCountOut:
    """How many members this rule matches right now.

    The product is optional and only scopes the coverage filters
    (``current_plan_codes`` / ``coverage_state``) — an unknown code leaves them
    inert rather than 404-ing, because the count is a live readout the picker
    fires on every keystroke and must never be the thing that breaks a page.
    """
    product = (
        resolve_product_by_code(db, py, body.product_code) if body.product_code else None
    )
    selection = svc.resolve_selection(
        db, py, body.query, product_id=product.id if product else None
    )
    return MemberQueryCountOut(
        total=len(selection.employees),
        unresolved=[u.out() for u in selection.unresolved],
    )


@router.post(
    "/policy-years/{policy_year_id}/member-query/resolve",
    response_model=MemberListResolveOut,
)
def resolve_member_list(
    policy_year_id: str,
    body: MemberListResolveIn,
    py: PolicyYear = Depends(load_policy_year),
    db: Session = Depends(get_db),
) -> MemberListResolveOut:
    """Turn pasted text (a spreadsheet column, an email list) into members.

    Unmatched entries come back so they can be shown IN the picker — the old
    flow only discovered a bad staff id after a full preview run.
    """
    matched, unmatched, duplicates = svc.resolve_member_list(
        db, py, body.text, include_terminated=body.include_terminated
    )
    return MemberListResolveOut(
        matched=matched, unmatched=unmatched, duplicates=duplicates
    )

File: frontend/src/api/memberQuery.ts
/**
 * Roster selection — the vocabulary, the live headcount, and pasted-list
 * resolution behind the member picker.
 *
 * A selection is a RULE (`MemberQuery`), not a list of people. The same object
 * goes to the count endpoint while the broker is composing it and to the bulk
 * preview/apply afterwards, so what they were shown and what runs are the same
 * query.
 */
import { useMutation, useQuery } from "@tanstack/react-query";
import { api } from "@/api/client";
import { useSession } from "@/stores/session";
import type { EmployeeList } from "@/types";

export type CoverageState = "any" | "default" | "overridden" | "declined";
export type MatchStatus = "any" | "matched" | "unmatched";

export interface AttributeFilter {
  key: string;
  values: string[];
  op?: "in" | "not_in";
}

export interface MemberQuery {
  q?: string | null;
  include_terminated?: boolean;
  category_ids?: string[];
  product_codes?: string[];
  current_plan_codes?: string[];
  coverage_state?: CoverageState;
  /** Whether roster matching found the member a cohort. */
  match_status?: MatchStatus;
  attributes?: AttributeFilter[];
  age?: { min?: number | null; max?: number | null } | null;
  /** Explicit additions (ticked rows, a pasted list). */
  employee_ids?: string[];
  staff_ids?: string[];
  /** Explicit removals, applied last — how unticking a row is expressed. */
  exclude_employee_ids?: string[];
}

export interface FacetValue {
  value: string;
  count: number;
}

export interface AttributeFacet {
  key: string;
  label: string;
  values: FacetValue[];
  truncated: boolean;
}

export interface CategoryFacet {
  id: string;
  label: string;
  product_code: string | null;
  count: number;
}

export interface PlanFacet {
  code: string;
  count: number;
}

export interface ProductFacet {
  id: string;
  code: string;
  name: string | null;
  covered: number;
  declined: number;
  plans: PlanFacet[];
}

export interface MemberFacets {
  employees_total: number;
  terminated_total: number;
  attributes: AttributeFacet[];
  categories: CategoryFacet[];
  products: ProductFacet[];
}

export interface UnresolvedRef {
  kind: "employee_id" | "staff_id";
  value: string;
  reason: string;
}

export interface MemberQueryCount {
  total: number;
  unresolved: UnresolvedRef[];
}

export interface ResolvedMember {
  id: string;
  staff_id: string;
  employee_name: string | null;
  matched_on: "staff_id" | "nric";
}

export interface MemberListResolve {
  matched: ResolvedMember[];
  unmatched: string[];
  duplicates: number;
}

/** Does this query select anybody at all? Mirrors the server's validator, so the
 *  picker can skip a request that would only ever come back 422. */
export function isQueryEmpty(query: MemberQuery): boolean {
  return !(
    query.q ||
    query.category_ids?.length ||
    query.product_codes?.length ||
    query.current_plan_codes?.length ||
    query.attributes?.length ||
    query.age ||
    (query.coverage_state && query.coverage_state !== "any") ||
    (query.match_status && query.match_status !== "any") ||
    query.employee_ids?.length ||
    query.staff_ids?.length
  );
}

/**
 * A filtered page of the roster.
 *
 * POST-for-read because `attributes[]` and `age` are nested and don't survive a
 * query string — the count and resolve endpoints are POST for the same reason.
 * Unlike `useMemberQueryCount` this accepts an EMPTY query: for a listing that
 * is the default view, not a footgun.
 */
export function useMemberQueryList(
  policyYearId: string | undefined,
  query: MemberQuery,
  page: { offset: number; limit: number },
  productCode?: string,
) {
  const clientId = useSession((s) => s.activeClientId);
  return useQuery({
    // Prefixed "employees" so every existing
    // `invalidateQueries({queryKey: ["employees"]})` in api/hooks.ts also
    // refreshes this listing. Without the prefix the table silently kept
    // showing stale rows after a save, a delete or a matching run.
    queryKey: [
      "employees",
      "query",
      clientId,
      policyYearId,
      productCode ?? null,
      page.offset,
      page.limit,
      query,
    ],
    queryFn: () =>
      api.post<EmployeeList>(
        `/policy-years/${policyYearId}/member-query/list`,
        { query, product_code: productCode ?? null, ...page },
      ),
    enabled: !!policyYearId,
    // Keep the previous page on screen while the next one loads, so changing a
    // filter doesn't blank the table under the broker.
    placeholderData: (prev) => prev,
  });
}

export function useMemberFacets(policyYearId: string | undefined) {
  const clientId = useSession((s) => s.activeClientId);
  return useQuery({
    // The active client is in the key: every tenant-scoped query must be, or a
    // company switch serves the previous company's roster from cache.
    queryKey: ["employees", "facets", clientId, policyYearId],
    queryFn: () =>
      api.get<MemberFacets>(`/policy-years/${policyYearId}/member-facets`),
    enabled: !!policyYearId,
  });
}

/** Live "N members match", refreshed as the filters change. */
export function useMemberQueryCount(
  policyYearId: string | undefined,
  query: MemberQuery,
  productCode: string | undefined,
) {
  const clientId = useSession((s) => s.activeClientId);
  const enabled = !!policyYearId && !isQueryEmpty(query);
  return useQuery({
    queryKey: ["employees", "count", clientId, policyYearId, productCode, query],
    queryFn: () =>
      api.post<MemberQueryCount>(
        `/policy-years/${policyYearId}/member-query/count`,
        { query, product_code: productCode ?? null },
      ),
    enabled,
    // The headcount is a readout, not state to act on — keep the previous number
    // on screen while the next one loads instead of flashing a skeleton on every
    // keystroke.
    placeholderData: (prev) => prev,
  });
}

export function useResolveMemberList(policyYearId: string | undefined) {
  return useMutation({
    mutationFn: (body: { text: string; include_terminated?: boolean }) =>
      api.post<MemberListResolve>(
        `/policy-years/${policyYearId}/member-query/resolve`,
        body,
      ),
  });
}
